### PR TITLE
[Enhancement]eliminate agg when group by unique key

### PR DESCRIPTION
**Pull request by liubotao**
## Why I'm doing:

When a column used in a SQL query's Group By statement has a unique attribute, aggregation can be eliminated,
and the LogicalAggregationOperator can be replaced with a LogicalProjectOperator.

## What I'm doing:

eliminate agg when group by unique key

transform sql demo:

```
1. select count(row1) from demo group by pk -> SELECT IF(row1 IS NULL, 0, 1) from demo;
2. select avg(row1) from demo group by pk -> SELECT row1 FROM demo;
```

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.3
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/EliminateAggRule.java [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+// example:
+// 1. select count(row1) from demo group by pk -> SELECT IF(row1 IS NULL, 0, 1) from demo;
+// 2. select avg(row1) from demo group by pk -> SELECT row1 FROM demo;
+//
+
+public class EliminateAggRule extends TransformationRule {
+
+    private EliminateAggRule() {
+        super(RuleType.TF_ELIMINATE_AGG, Pattern.create(OperatorType.LOGICAL_AGGR)
+                .addChildren(Pattern.create(OperatorType.PATTERN_LEAF)));
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/EliminateAggRule.java:68]: do we need to check the agg Children pattern?  such as  scan-> left join-> agg.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/EliminateAggRule.java [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+    @Override
+    public boolean check(OptExpression input, OptimizerContext context) {
+        LogicalAggregationOperator aggOp = input.getOp().cast();
+        if (aggOp.getGroupingKeys().isEmpty()) {
+            return false;
+        }
+
+        for (Map.Entry<ColumnRefOperator, CallOperator> entry : aggOp.getAggregations().entrySet()) {
+            String fnName = entry.getValue().getFnName();
+            if (!(fnName.equals(FunctionSet.SUM) || fnName.equals(FunctionSet.COUNT) ||
```
Comment (packy92) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/EliminateAggRule.java:86]: build to supported functionName set to check.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/EliminateAggRule.java (hunk 1 above)
Comment (liubotao) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/EliminateAggRule.java:68]: If the agg child node can output the unique key attribute, can remove agg

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/EliminateAggRule.java [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+            }
+        }
+
+        // collect uk pk key
+        UKFKConstraintsCollector collector = new UKFKConstraintsCollector();
+        input.getOp().accept(collector, input, null);
+
+        OptExpression childOptExpression = input.inputAt(0);
+        for (ColumnRefOperator columnRefOperator : aggOp.getGroupingKeys()) {
+            if (childOptExpression.getConstraints().getUniqueConstraint(columnRefOperator.getId()) == null) {
```
Comment (packy92) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/EliminateAggRule.java:101]: the rule only takes effect only when group by keys contain all unique key. If group by key is (colA, colB) but unique key is (colA, colB, colC), we cannot remove the agg.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/EliminateAggRule.java [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+        LogicalProjectOperator newProjectOp = LogicalProjectOperator.builder().setColumnRefMap(newProjectMap).build();
+        return List.of(OptExpression.create(newProjectOp, input.inputAt(0).getInputs()));
+    }
+
+    private ScalarOperator handleAggregationFunction(String fnName, CallOperator callOperator) {
+        if (fnName.equals(FunctionSet.COUNT)) {
+            return rewriteCountFunction(callOperator);
+        } else if (fnName.equals(FunctionSet.SUM) || fnName.equals(FunctionSet.AVG) ||
+                fnName.equals(FunctionSet.FIRST_VALUE) || fnName.equals(FunctionSet.MAX) ||
+                fnName.equals(FunctionSet.MIN) || fnName.equals(FunctionSet.GROUP_CONCAT)) {
```
Comment (packy92) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/EliminateAggRule.java:132]: the return type of sum/avg may different from its arg type. need add cast if necessary.

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/EliminateAggRule.java (hunk 2 above)
Comment (packy92) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/EliminateAggRule.java:86]: check the distinct flag in agg function.

**Review by packy92 — APPROVED** [review]

**Review by kangkaisen — APPROVED** [review]

**Review by Seaven — APPROVED** [review]

**Pull request merged**